### ISSUE TITLE
CI: Flaky tests on slow runners

### ISSUE THREAD

**Issue by OliverO2** [kotest/kotest#4113]
**:kotest-assertions:kotest-assertions-core:jvmTest**
* [x] com.sksamuel.kotest.assertions.timing.ContinuallyTest[jvm] (macos-latest)
  ```
  continually should[jvm]
      Test fail tests start off as passing then fail within the period[jvm] FAILED (3s)

      java.lang.AssertionError: Expected exception kotlin.Throwable but no exception was thrown.
  ```

  In progress: https://github.com/kotest/kotest/pull/4109

* [x] com.sksamuel.kotest.matchers.concurrent.suspension.ConcurrentTest[jvm] (windows-latest)
    ```
    Test should fail when given lambda pass before the given time range[jvm] FAILED (1.2s)
    
    java.lang.AssertionError: Expected exception java.lang.AssertionError but no exception was thrown.
    ```
  In progress: https://github.com/kotest/kotest/pull/4115

* [x] io.kotest.assertions.nondeterministic.UntilTest[jvm] (macos-latest)
    ```
    Test until with config[jvm] FAILED (2.2s)
  
    java.lang.AssertionError: Block failed after 2s; attempted 36 time(s)
    The first error was caused by: expected:<true> but was:<false>
    ```

  In progress: https://github.com/kotest/kotest/pull/4132

- [x] io.kotest.assertions.nondeterministic.EventuallyTest[jvm]

  ```  
    Test handle java assertion errors[jvm] FAILED
  
    java.lang.AssertionError: Block failed after 157ms; attempted 1 time(s)
    The first error was caused by: boom
    java.lang.AssertionError: boom
  ```

  In progress: #4114

**:kotest-assertions:kotest-assertions-shared:jvmTest**
* [x] com.sksamuel.kotest.assertions.until.UntilTest[jvm] (macos-latest)
    ```
    Test until with boolean predicate and interval[jvm] FAILED (2s)
    
    java.lang.AssertionError: Until block failed after 2.049s; attempted 36 time(s); FixedInterval(duration=10ms) delay between attempts
    Test until with patience config[jvm] FAILED (2s)
    
    java.lang.AssertionError: Until block failed after 2.016s; attempted 35 time(s); FixedInterval(duration=10ms) delay between attempts
    Test until with predicate and interval[jvm] FAILED
    
    java.lang.AssertionError: 115ms should be < 100ms
    ```

  In progress: https://github.com/kotest/kotest/pull/4132

* [x] io.kotest.assertions.CluesTests[jvm]
    ```
    Test concurrent withClue invocations should be isolated from each other[jvm] FAILED (1.4s)
    
    java.lang.AssertionError: 46 should be >= 50
    ```

  In progress: https://github.com/kotest/kotest/pull/4132

* [x] com.sksamuel.kotest.assertions.async.TimeoutTest[jvm]
    ``` 
    Test shouldTimeout should fail if a coroutine does not timeout[jvm] FAILED
    java.lang.AssertionError: Expected exception java.lang.AssertionError but no exception was thrown.
    ```

  In progress: https://github.com/kotest/kotest/pull/4132

**:kotest-framework:kotest-framework-concurrency:jvmTest**
* [x] io.kotest.framework.concurrency.UntilSpec[jvm] (macos-latest, windows-latest)
    ```
    Test until with boolean predicate and interval[jvm] FAILED (2.1s)
    
    java.lang.AssertionError: Eventually block failed after 2000ms; attempted 40 time(s); FixedInterval(duration=10) delay between attempts
    The provided predicate failed 40 times
    
    Test until with predicate and interval[jvm] FAILED
    
    java.lang.AssertionError: 174 should be < 100
    ```

  In progress: https://github.com/kotest/kotest/pull/4132

* [x] io.kotest.framework.concurrency.EventuallySpec[jvm]
    ``` 
    Test eventually handles java assertion errors[jvm] FAILED
  
    java.lang.AssertionError: Eventually block failed after 100ms; attempted 1 time(s); FixedInterval(duration=25) delay between attempts
    The first error was caused by: boom
    java.lang.AssertionError: boom
    ```

  In progress: https://github.com/kotest/kotest/pull/4142
    - also: https://github.com/kotest/kotest/pull/4149

* [x] io.kotest.framework.concurrency.ContinuallySpec[jvm]
    ``` 
    Test continually fails tests that start off as passing then fail before the time is up[jvm] FAILED (1.1s)
    400.milliseconds
    400.milliseconds  java.lang.AssertionError: Throwable should match regex: Test failed after d+(.d*)?ms; expected to pass for 12h; attempted 10 times
    400.milliseconds  Underlying failure was: expected:<true> but was:<false>
    400.milliseconds  Actual was:
    400.milliseconds  "Test failed after 1.101072583s; expected to pass for 12h; attempted 10 times
    400.milliseconds  Underlying failure was: expected:<true> but was:<false>"
    ```

  In progress: https://github.com/kotest/kotest/pull/4142

**:kotest-framework:kotest-framework-engine:jvmTest**
* [x] com.sksamuel.kotest.engine.test.timeout.EngineTimeoutTest[jvm] (windows-latest)
    ```
    Test timeouts should be applied by the engine to suspend delays[jvm] FAILED (3.9s)
    
    io.kotest.assertions.AssertionFailedError: expected:<"Test 'a' did not complete within 1ms"> but was:<"Timed out waiting for 1 ms">
    ```

  In progress: https://github.com/kotest/kotest/pull/4124
    - superseded by: https://github.com/kotest/kotest/pull/4149

- [x] com.sksamuel.kotest.engine.test.timeout.GlobalTimeoutTest[jvm]
  ```
  Test global timeouts should apply if no other timeout is set[jvm] FAILED (2.7s)

  java.lang.AssertionError: Expected true but actual was null

  Caused by: java.lang.AssertionError: Expected true but actual was null
  ```
  https://github.com/kotest/kotest/actions/runs/9555635058/job/26339325763?pr=4114#step:7:3157

  In progress: https://github.com/kotest/kotest/pull/4124
    - superseded by: https://github.com/kotest/kotest/pull/4149

* [x] com.sksamuel.kotest.engine.test.timeout.MultipleTestTimeoutTest[jvm] (macos-latest)
    ```
    Test Test 5[jvm] FAILED
    
    kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 40 ms
    
    Test Test 7[jvm] FAILED
    
    kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 40 ms
    ```

  In progress: https://github.com/kotest/kotest/pull/4132

* [x] com.sksamuel.kotest.engine.test.timeout.SpecInvocationTimeoutShouldNotApplyToContainersTest[jvm] (macos-latest, windows-latest)
    ```
    invocation timeouts at the spec level should not be applied to containers[jvm]

        Test suspending inner test[jvm] FAILED
    
        io.kotest.engine.test.interceptors.TestTimeoutException: Test 'suspending inner test' did not complete within 50ms
    
        Test blocking inner test[jvm] FAILED
    
        io.kotest.engine.test.interceptors.TestTimeoutException: Test 'blocking inner test' did not complete within 50ms
    ```

  In progress: https://github.com/kotest/kotest/pull/4132
    - also: https://github.com/kotest/kotest/pull/4149

* [x] io.kotest.engine.concurrency.TimeoutTest[jvm]

    ```
    Test detection with blocking job[jvm] FAILED
    java.lang.AssertionError: Expected exception kotlinx.coroutines.TimeoutCancellationException but no exception was thrown.
    ```

  In progress: https://github.com/kotest/kotest/pull/4132
    - also: https://github.com/kotest/kotest/pull/4149

**:kotest-property:jvmTest**
* [x] com.sksamuel.kotest.property.PropTestConfigConstraintsTest[jvm] (windows-latest)
    ```
    Test PropTestConfig constraints should support durations[jvm] FAILED
    
    java.lang.AssertionError: 420 should be < 400
    ```

  In progress: https://github.com/kotest/kotest/pull/4132

**:kotest-runner:kotest-runner-junit5:jvmTest**
* [x] com.sksamuel.kotest.runner.junit5.PerformanceTest[jvm]
    ```
    Test performance of multiple tests[jvm] FAILED (20.8s)
  
    kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 20000 ms
    ```

  In progress: https://github.com/kotest/kotest/pull/4132

#### `:kotest-property:jvmTest` `Test Shrinks all components to minimum value[jvm] FAILED`

- [ ] https://github.com/kotest/kotest/actions/runs/9864899341/job/27240897776?pr=4141#step:7:976
```
> Task :kotest-property:jvmTest
com.***.kotest.property.arbitrary.BindShrinkTest[jvm]

  Test Shrinks all components to minimum value[jvm] FAILED

  java.lang.AssertionError: Test failed after 0s; attempted 5 times; underlying cause was "Property test failed for inputs
  
  0) MaximumComponents(a=1000, b=1000, c=1000, d=1000, e=1000, f=1000, g=1000, h=1000, i=1000, j=1000, k=1000, l=1000, m=1000, n=1000)
  
  Caused by java.lang.AssertionError: 1000 should be < 100 at
  	com.***.kotest.property.arbitrary.BindShrinkTest$1$2$2$stdout$1$1$1.invokeSuspend(BindShrinkTest.kt:54)
  	com.***.kotest.property.arbitrary.BindShrinkTest$1$2$2$stdout$1$1$1.invoke(BindShrinkTest.kt)
  	com.***.kotest.property.arbitrary.BindShrinkTest$1$2$2$stdout$1$1$1.invoke(BindShrinkTest.kt)
  	io.kotest.property.internal.ProptestKt$proptest$3$2.invokeSuspend(proptest.kt:49)
  
  " should include substring "Shrink result (after 45 shrinks) => MaximumComponents(a=0, b=0, c=0, d=0, e=0, f=0, g=0, h=0, i=0, j=0, k=0, l=0, m=100, n=0)"

  Caused by: java.lang.AssertionError: Test failed after 0s; attempted 5 times; underlying cause was "Property test failed for inputs
  
  0) MaximumComponents(a=1000, b=1000, c=1000, d=1000, e=1000, f=1000, g=1000, h=1000, i=1000, j=1000, k=1000, l=1000, m=1000, n=1000)
  
  Caused by java.lang.AssertionError: 1000 should be < 100 at
  	com.***.kotest.property.arbitrary.BindShrinkTest$1$2$2$stdout$1$1$1.invokeSuspend(BindShrinkTest.kt:54)
  	com.***.kotest.property.arbitrary.BindShrinkTest$1$2$2$stdout$1$1$1.invoke(BindShrinkTest.kt)
  	com.***.kotest.property.arbitrary.BindShrinkTest$1$2$2$stdout$1$1$1.invoke(BindShrinkTest.kt)
  	io.kotest.property.internal.ProptestKt$proptest$3$2.invokeSuspend(proptest.kt:49)
  
  " should include substring "Shrink result (after 45 shrinks) => MaximumComponents(a=0, b=0, c=0, d=0, e=0, f=0, g=0, h=0, i=0, j=0, k=0, l=0, m=100, n=0)"

  Caused by: java.lang.AssertionError: "Property test failed for inputs
  
  0) MaximumComponents(a=[1000](https://github.com/kotest/kotest/actions/runs/9864899341/job/27240897776?pr=4141#step:7:1001), b=1000, c=1000, d=1000, e=1000, f=1000, g=1000, h=1000, i=1000, j=1000, k=1000, l=1000, m=1000, n=1000)
  
  Caused by java.lang.AssertionError: 1000 should be < 100 at
  	com.***.kotest.property.arbitrary.BindShrinkTest$1$2$2$stdout$1$1$1.invokeSuspend(BindShrinkTest.kt:54)
  	com.***.kotest.property.arbitrary.BindShrinkTest$1$2$2$stdout$1$1$1.invoke(BindShrinkTest.kt)
  	com.***.kotest.property.arbitrary.BindShrinkTest$1$2$2$stdout$1$1$1.invoke(BindShrinkTest.kt)
  	io.kotest.property.internal.ProptestKt$proptest$3$2.invokeSuspend(proptest.kt:49)
  
  " should include substring "Shrink result (after 45 shrinks) => MaximumComponents(a=0, b=0, c=0, d=0, e=0, f=0, g=0, h=0, i=0, j=0, k=0, l=0, m=100, n=0)"
```

In progress: https://github.com/kotest/kotest/pull/4167

### DiscoveryTestWithoutSelectors - OutOfMemoryError: Java heap space

https://github.com/kotest/kotest/actions/runs/9998565996/job/27637892318?pr=4156#step:7:3384

https://scans.gradle.com/s/thygkik33v3r2

```
> Task :kotest-runner:kotest-runner-junit5:jvmTest
com.***.kotest.runner.junit5.DiscoveryTestWithoutSelectors[jvm]
  Test kotest should recognize suffixed class name pattern filters[jvm] FAILED (1.8s)
  io.kotest.engine.config.ConfigurationException: io.kotest.engine.config.ConfigurationException: io.github.classgraph.ClassGraphException: Uncaught exception during scan
  Caused by: io.kotest.engine.config.ConfigurationException: io.github.classgraph.ClassGraphException: Uncaught exception during scan
  Caused by: io.github.classgraph.ClassGraphException: Uncaught exception during scan
  Caused by: java.lang.OutOfMemoryError: Java heap space

FAILURE: Executed 106 tests in 4m 40s (1 failed, 1 skipped)
```

### `com.sksamuel.kotest.engine.spec.examples.DescribeSpecExample[jvm]`

https://github.com/kotest/kotest/actions/runs/10133804332/job/28019385046?pr=4218#step:8:2652

```
> Task :kotest-framework:kotest-framework-engine:jvmMaxJdkTest
com.sksamuel.kotest.engine.spec.examples.DescribeSpecExample[jvm]
  a describe with config[jvm]
    with some describe[jvm]
      with some context[jvm]
        Test executionError[jvm] FAILED
        kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 234 ms
        Caused by: kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 234 ms

FAILURE: Executed 1921 tests in 13m 16s (1 failed)
1921 tests completed, 1 failed
```

- [ ] https://github.com/kotest/kotest/pull/4232

### `com.sksamuel.kotest.engine.test.TestCaseSourceRefTest[jvm]`

https://github.com/kotest/kotest/actions/runs/10138373663/job/28033374831?pr=4221#step:8:2174

https://scans.gradle.com/s/c3ygboydkzmis

```
com.sksamuel.kotest.engine.test.TestCaseSourceRefTest[jvm]
  Test source ref should be performant[jvm] FAILED
  java.lang.AssertionError: 2m 3.600324900s should be < 2m
  Caused by: java.lang.AssertionError: 2m 3.600324900s should be < 2m
```

- [ ] Up for grabs

**Comment by OliverO2** [comment]
@aSemy Maybe you could consider some of the above for #4105 and select the addressed tests in the above list. Let me know if I should fix any of those remaining.

**Comment by aSemy** [comment]
Nice! Thanks for making the list.

ContinuallyTest should be fixed by #4109, which uses kotlinx.coroutines virtual time. I expect that the other tests could be fixed in the same way. 

I'm also just finishing up a PR for fixing EventuallyTest (it's not on the list above but I've seen it fail in one of the test runs).

Feel free to pick any of the others - let's coordinate in this issue to make sure we don't overlap?

**Comment by OliverO2** [comment]
Great, let's do it this way! I'll call it a day for now but if I'll pick stuff, I'll leave a note here. And if I merge with positive CI results, I'll select the test in the above TODO list. Feel free to edit the above list, adding additional failing tests as required.

**Comment by OliverO2** [comment]
I just had a timeout on a local build with https://github.com/kotest/kotest/blob/0ec32b335362a98390285c6067c2d65f8e2ab581/kotest-runner/kotest-runner-junit5/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/PerformanceTest.kt#L12

This was seemingly caused by a parallel run of `:kotest-framework:kotest-framework-multiplatform-plugin-gradle:check`, which is very resource intensive as it fires up secondary Gradle builds.

Since #4076 we have `org.gradle.parallel=true`. That means we could always run into unfortunate interactions between parallel tasks which might distort timing in a way that is almost impossible to predict. Within a single Gradle test task, we can annotate time-sensitive tests with `@DoNotParallelize`. This, however, will not work across tasks.

We have some real time performance expectations in tests which will cause flakiness forever when used in conjunction with parallel testing. So we should probably either
* avoid parallel processing for test runs altogether (we could still do parallel builds), or
* group tests according to their time sensitivity and execute the sensitive set with parallel processing disabled.

What do you think?

**Comment by aSemy** [comment]
I think that when configuration cache is enabled `org.gradle.parallel` has no effect. Tasks from multiple projects will run in parallel.

There are a few performance tests scattered about. TBH, I think they should be split out into dedicated performance tests using JMH. However, that could be quite a lot of work.

Alternatively, as a quicker fix, we could tag performance tests and run them separately to the standard tests, with `--no-configuration-cache`?

**Comment by OliverO2** [comment]
OK, seems like there is always the exception to the exception in Gradle 🙃. The tag idea sounds promising. I'll create a separate issue for that.

**Comment by aSemy** [comment]
Quite a lot of the engine tests use `TestEngineLauncher#launch`, which uses `runBlocking {}`. I suspect this would be a cause of flaky tests, because the tests won't be able to wait for the result of the tests.

I wonder if using `TestEngineLauncher#async` would be better.

https://github.com/kotest/kotest/blob/2a619504a4d01075a2e42c61ac1eb7f8f3d83bdd/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/TestEngineLauncher.kt#L238-L257

**Comment by OliverO2** [comment]
Well, `async` is suspending, so the border crossing between blocking and suspending worlds happens just at some layer above, where you'll always find some `runBlocking`. The exception is JS, where the outer layer must be a `Promise` which the JS test frameworks expect to wait for test results.

`runBlocking` is a delicate beast because you can block your main thread if you're not careful. But such blocking will most probably reproduce constantly.

**Comment by aSemy** [comment]
Hmmm yes, that makes sense. I was asking because the GlobalTimeoutTest failure (added to the issue above) above indicates that the `global timeouts should apply if no other timeout is set` test doesn't wait for the launched tests to complete. My thought was that `runBlocking {}` is launching a new thread...

**Comment by OliverO2** [comment]
I can try to look at GlobalTimeoutTest tomorrow. `runBlocking` does nothing with threads, just happily runs on the current one. It basically starts an event loop on which it schedules coroutines.

**Comment by aSemy** [comment]
Another idea: tag add `@Tag("Deprecated")` to all tests of deprecated functionality and run them in a separate GitHub step. The deprecated tests are often more flaky, and we can either treat them as less important (don't mark the step as mandatory), or add flags to modify the timings (because there are fewer deprecated tests it's okay if they don't run in parallel).

**Comment by OliverO2** [comment]
Yes, that's also a good idea.

**Comment by OliverO2** [comment]
Seems like we cannot proceed with tag-based exclusion due to #4135.

**Comment by aSemy** [comment]
> Seems like we cannot proceed with tag-based exclusion due to #4135.

Silver lining: It is good that we are finding bugs 🙃

**Comment by OliverO2** [comment]
Yep. I guess we're just completing a phase with the largest amount of testing Kotest has seen so far.

**Comment by OliverO2** [comment]
Regarding `GlobalTimeoutTest`: I can't see why `TestEngineLauncher.launch()` would not wait until all tests are finished. The `runBlocking`, which it invokes, does. So unless some inner coroutine breaks out of structured concurrency, everything should be fine even without your changes. However, at least your PR provides a viable workaround.

**Comment by aSemy** [comment]
Hmmm it looks like my workaround for GlobalTimeoutTest doesn't work reliably. There's a failure on Windows:

https://github.com/kotest/kotest/actions/runs/9624071919/job/26547346172?pr=4141#step:7:2206

```
> Task :kotest-framework:kotest-framework-engine:jvmTest
com.sksamuel.kotest.engine.test.timeout.GlobalTimeoutTest[jvm]
  Test global timeouts should apply if no other timeout is set[jvm] FAILED (9.2s)
  io.kotest.assertions.AssertionFailedError: Collection should contain exactly: ["blocked", "suspend"] but was: ["suspend"]
  Some elements were missing: ["blocked"]
  expected:<["blocked", "suspend"]> but was:<["suspend"]>
```

**Comment by OliverO2** [comment]
Testing a new hypothesis for `GlobalTimeoutTest`: The 4 ms timeout was so short that one or both tests were interrupted before launch.
